src/api/pubsub.js

```diff
@@ -0,0 +1,109 @@
+'use strict'
+
+const promisify = require('promisify-es6')
+const bs58 = require('bs58')
+const Base64 = require('js-base64').Base64
+const Stream = require('stream')
+const Readable = Stream.Readable
+const http = require('http')
+
+let activeSubscriptions = []
+
+const subscriptionExists = (subscriptions, topic) => {
+  return subscriptions.includes(topic)
+}
+const removeSubscription = (subscriptions, topic) => {
+  const indexToRemove = subscriptions.indexOf(topic)
+  return subscriptions.filter((el, index) => {
+    return index !== indexToRemove
+  })
+}
+const addSubscription = (subscriptions, topic) => {
+  return subscriptions.concat([topic])
+}
+const parseMessage = (message) => {
+  return Object.assign({}, message, {
+    from: bs58.encode(message.from),
+    data: Base64.decode(message.data),
+    seqno: Base64.decode(message.seqno)
+  })
+}
+
+module.exports = (send, config) => {
+  return {
+    subscribe: promisify((topic, options, callback) => {
+      if (typeof options === 'function') {
+        callback = options
+        options = {}
+      }
+      if (!options) {
+        options = {}
+      }
+
+      var rs = new Readable({objectMode: true})
+      rs._read = () => {}
+
+      if (!subscriptionExists(activeSubscriptions, topic)) {
+        activeSubscriptions = addSubscription(activeSubscriptions, topic)
+      } else {
+        return callback(new Error('Already subscribed to ' + topic), null)
+      }
+
+      let url = '/api/v0/pubsub/sub/' + topic
+      if (options.discover) {
+        url = url + '?discover=true'
```

Comment:
Sure, it can but the value given by it is less :) I would rather refactor everything to use the request-api instead of it's own `http.get`, but would require refactor of request-api as well and then your ongoing fetch changes will conflict. 

I would say we keep this as is now, and change it in the future when needed, once the API stabilized. Keep in mind, pub/sub is not even released in go-ipfs yet, and usage of this will be minimal.